framework/include/utils/SymmetricRankFourTensor.h

```diff
@@ -0,0 +1,577 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "Moose.h"
+#include "ADRankTwoTensorForward.h"
+#include "ADRankFourTensorForward.h"
+#include "ADSymmetricRankTwoTensorForward.h"
+#include "ADSymmetricRankFourTensorForward.h"
+#include "MooseUtils.h"
+#include "MathUtils.h"
+
+#include "libmesh/libmesh.h"
+#include "libmesh/tuple_of.h"
+#include "libmesh/int_range.h"
+
+#include "metaphysicl/raw_type.h"
+
+#include <petscsys.h>
+
+// Eigen includes
+#include <Eigen/Core>
+#include <Eigen/Dense>
+
+#include <array>
+
+using libMesh::Real;
+using libMesh::tuple_of;
+namespace libMesh
+{
+template <typename>
+class TensorValue;
+template <typename>
+class TypeTensor;
+template <typename>
+class VectorValue;
+}
+
+// Forward declarations
+class MooseEnum;
+
+namespace MathUtils
+{
+/**
+ * Helper function template specialization to set an object to zero.
+ * Needed by DerivativeMaterialInterface
+ */
+template <>
+void mooseSetToZero<SymmetricRankFourTensor>(SymmetricRankFourTensor & v);
+template <>
+void mooseSetToZero<ADSymmetricRankFourTensor>(ADSymmetricRankFourTensor & v);
+}
+
+/**
+ * SymmetricRankFourTensorTempl is designed to handle a symmetric N-dimensional fourth order tensor,
+ * C. Since N is hard-coded to 3, SymmetricRankFourTensorTempl holds 36 separate C_ij entries.
+ * Within the code i,j = 0, .., 5.
+ */
+template <typename T>
+class SymmetricRankFourTensorTempl
+{
+public:
+  ///@{ tensor dimension, Mandel matrix dimension, and Mandel matrix size
+  static constexpr unsigned int Ndim = 3;
+  static constexpr unsigned int N = 2 * Ndim;
+  static constexpr unsigned int N2 = N * N;
+  ///@}
+
+  /// returns the 1, sqrt(2), or 2 prefactor in the Mandel notation for the indices i,j ranging from 0-5.
+  static constexpr Real mandelFactor(unsigned int i, unsigned int j)
+  {
+    return i < 3 ? (j < 3 ? 1.0 : MathUtils::sqrt2) : (j < 3 ? MathUtils::sqrt2 : 2.0);
```

Comment:
I suppose I should use `Ndim` here to not come off as a hypocrite...

Reply:
Why the capitalization though? I only capitalize named constructors.

Reply:
I may have totally misunderstood it. I thought the convention of method naming was camelCase unless "camel" is a significant word, in which case we still write "CamelCase". I might be wrong.

Speaking of named constructors, I thought the first letter should be lower case since ultimately it's a method not a constructor. But I also see why you are capitalizing it. I guess I'm just not sure what the convention was.

Reply:
This is an example I found from my c++ textbook:
```cpp
class Point {
public:
  static Point rectangular(float x, float y);      // Rectangular coordindates
  static Point polar(float radius, float angle);   // Polar coordinates

  ...
private:
  Point();
  Point(float x, float y);     // Rectangular coordinates
  float x_, y_;
};
```
but maybe MOOSE is using a different naming convention.